### PR TITLE
[Shallow] clean deprecations

### PR DESCRIPTION
**Pull request by miguelmaso**
**Description**
Fix some deprecations

**Changelog**
- Port some parallel loops to parallel utilities
- Update FindGlobalNodalNeighboursProcess constructor
- Remove unused method

### CODE REVIEW

File: applications/ShallowWaterApplication/custom_processes/apply_sinusoidal_function_process.cpp [hunk 1]
```diff
@@ -67,28 +68,22 @@ void ApplySinusoidalFunctionProcess<TVarType>::ExecuteInitializeSolutionStep()
...
-    #pragma omp parallel for
-    for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++)
-    {
-        auto it_node = mrModelPart.NodesBegin() + i;
-        array_1d<double, 3> direction = it_node->FastGetSolutionStepValue(NORMAL);
-        it_node->FastGetSolutionStepValue(mrVariable) = -modulus * direction;
-    }
+    block_for_each(mrModelPart.Nodes(), [&](NodeType& rNode){
+        array_1d<double, 3> direction = rNode.FastGetSolutionStepValue(NORMAL);
+        rNode.FastGetSolutionStepValue(mrVariable) = -modulus * direction;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_processes/apply_sinusoidal_function_process.cpp]:
```suggestion
        noalias(rNode.FastGetSolutionStepValue(mrVariable)) = -modulus * direction;
```
This is faster for array-type variables

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 2]
```diff
@@ -321,34 +273,18 @@ class MoveShallowWaterParticleUtility
...
-
-        #pragma omp parallel for
-        for(int kkk=0; kkk<number_of_threads; kkk++)
-        {
-            for(unsigned int ii=element_partition[kkk]; ii<element_partition[kkk+1]; ii++)
-            {
-                ModelPart::ElementsContainerType::iterator ielem = ielembegin+ii;
-                Geometry<Node<3> >& geom = ielem->GetGeometry();
+        block_for_each(mrModelPart.Elements(), [&](Element& rElem){
+            GeometryType& geom = rElem.GetGeometry();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            const GeometryType& geom = rElem.GetGeometry();
```
(maybe I'm wrong)

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 3]
```diff
@@ -413,83 +329,38 @@ class MoveShallowWaterParticleUtility
...
-            for(unsigned int ii=node_partition[kkk]; ii<node_partition[kkk+1]; ii++)
-            {
-                ModelPart::NodesContainerType::iterator inode = inodebegin+ii;
-                inode->FastGetSolutionStepValue(DELTA_SCALAR1) = inode->FastGetSolutionStepValue(*mScalarVar1) - inode->FastGetSolutionStepValue(PROJECTED_SCALAR1);
-                inode->FastGetSolutionStepValue(DELTA_VECTOR1) = inode->FastGetSolutionStepValue(*mVectorVar1) - inode->FastGetSolutionStepValue(PROJECTED_VECTOR1);  //PROJECTED_VECTOR1
-            }
-        }
+        block_for_each(mrModelPart.Nodes(), [&](NodeType& rNode){
+            rNode.FastGetSolutionStepValue(DELTA_SCALAR1) = rNode.FastGetSolutionStepValue(*mScalarVar1) - rNode.FastGetSolutionStepValue(PROJECTED_SCALAR1);
+            rNode.FastGetSolutionStepValue(DELTA_VECTOR1) = rNode.FastGetSolutionStepValue(*mVectorVar1) - rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            noalias(rNode.FastGetSolutionStepValue(DELTA_VECTOR1)) = rNode.FastGetSolutionStepValue(*mVectorVar1) - rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1);
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 4]
```diff
@@ -523,108 +394,89 @@ class MoveShallowWaterParticleUtility
...
+        ResultContainerType results(max_results);
+        GlobalPointersVector<Element> elements_in_trajectory;
+        elements_in_trajectory.resize(20);
 
-            for(unsigned int ielem = element_partition[kkk]; ielem<element_partition[kkk+1]; ielem++)
-            {
-                ModelPart::ElementsContainerType::iterator old_element = ielembegin+ielem;
-                const int old_element_id = old_element->Id();
+        // block_for_each(mrModelPart.Elements(), tls, [&](Element& rElem, TLS& rTLS){
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
        auto i_elem_begin = mrModelPart.ElementsBegin();
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 5]
```diff
@@ -523,108 +394,89 @@ class MoveShallowWaterParticleUtility
...
+        #pragma omp parallel for firstprivate(results, elements_in_trajectory)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: The `firstprivate` clause can be emulated with the Thread Local Storage (TLS) of the `ParallelUtils`.

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 6]
```diff
@@ -649,144 +501,106 @@ class MoveShallowWaterParticleUtility
...
-                ModelPart::NodesContainerType::iterator inode = inodebegin+ii;
-                inode->FastGetSolutionStepValue(PROJECTED_SCALAR1)=0.0;
-                inode->FastGetSolutionStepValue(PROJECTED_VECTOR1)=ZeroVector(3);
-                inode->FastGetSolutionStepValue(YP)=0.0;
-            }
-        }
+
+        block_for_each(mrModelPart.Nodes(), [&](NodeType& rNode){
+            rNode.FastGetSolutionStepValue(PROJECTED_SCALAR1)=0.0;
+            rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1)=ZeroVector(3);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            noalias(rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1))=ZeroVector(3);
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 7]
```diff
@@ -649,144 +501,106 @@ class MoveShallowWaterParticleUtility
...
 
-                for (int iii=0; iii<number_of_particles_in_elem ; iii++ )
+                if (pparticle.GetEraseFlag() == false)
                 {
-                    if (iii==mMaxNumberOfParticles) // It means we are out of our portion of the array, abort loop!
-                        break;
-
-                    ShallowParticle& pparticle = element_particle_pointers[offset+iii];
+                    array_1d<double,3> & position = pparticle.Coordinates();
+                    const float& particle_scalar1 = pparticle.GetScalar1();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: Why using `float` and not `double`?

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 8]
```diff
@@ -649,144 +501,106 @@ class MoveShallowWaterParticleUtility
...
+                double & scalar = rNode.FastGetSolutionStepValue(PROJECTED_SCALAR1);
+                array_1d<double,3> & vector = rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1);
+                scalar /= sum_weights;
+                vector /= sum_weights;
             }
-        }
+            else // This should never happen because other ways to recover the information have been executed before, but leaving it just in case..
+            {
+                rNode.FastGetSolutionStepValue(PROJECTED_SCALAR1)=rNode.FastGetSolutionStepValue(*mScalarVar1,1);
+                rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1)=rNode.FastGetSolutionStepValue(*mVectorVar1,1);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
                noalias(rNode.FastGetSolutionStepValue(PROJECTED_VECTOR1))=rNode.FastGetSolutionStepValue(*mVectorVar1,1);
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 9]
```diff
@@ -804,46 +618,29 @@ class MoveShallowWaterParticleUtility
 
         const int offset = mOffset; //the array of pointers for each element has twice the required size so that we use a part in odd timesteps and the other in even ones.
                                     //(flag managed only by MoveParticles)
-        ModelPart::ElementsContainerType::iterator ielembegin = mrModelPart.ElementsBegin();
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
        auto i_elem_begin = mrModelPart.ElementsBegin();
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 10]
```diff
@@ -804,46 +618,29 @@ class MoveShallowWaterParticleUtility
...
+        IndexPartition<unsigned int>(mrModelPart.NumberOfElements()).for_each([&](unsigned int i){
+            ModelPart::ElementsContainerType::iterator ielem = i_elem_begin + i;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            auto ielem = i_elem_begin + i;
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 11]
```diff
@@ -868,92 +665,70 @@ class MoveShallowWaterParticleUtility
...
-            BoundedMatrix<double, (TDim+1), 3 > pos;
-            BoundedMatrix<double, (TDim+1) , (TDim+1) > N;
-            unsigned int freeparticle=0; //we start with the first position in the particles array
-
-            //int local_id=1;
-            for(unsigned int ii=elem_partition[k]; ii<elem_partition[k+1]; ii++)
-            {
-                //const int & elem_id = ielem->Id();
-                ModelPart::ElementsContainerType::iterator ielem = ielembegin+ii;
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
        auto i_elem_begin = mrModelPart.ElementsBegin();
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 12]
```diff
@@ -868,92 +665,70 @@ class MoveShallowWaterParticleUtility
...
+        #pragma omp parallel for firstprivate(results, pos, N, free_particle)
+        for (int ii = 0; ii < static_cast<int>(mrModelPart.NumberOfElements()); ++ii) {
+            ModelPart::ElementsContainerType::iterator ielem = i_elem_begin + ii;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            auto ielem = i_elem_begin + ii;
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 13]
```diff
@@ -868,92 +665,70 @@ class MoveShallowWaterParticleUtility
...
-            BoundedMatrix<double, (TDim+1) , (TDim+1) > N;
-            unsigned int freeparticle=0; //we start with the first position in the particles array
-
-            //int local_id=1;
-            for(unsigned int ii=elem_partition[k]; ii<elem_partition[k+1]; ii++)
-            {
-                //const int & elem_id = ielem->Id();
-                ModelPart::ElementsContainerType::iterator ielem = ielembegin+ii;
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
+        #pragma omp parallel for firstprivate(results, pos, N, free_particle)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: Same comment, you can use the TLS in here.

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 14]
```diff
@@ -868,92 +665,70 @@ class MoveShallowWaterParticleUtility
...
+                {
+                    bool keep_looking = true;
+                    while(keep_looking)
                     {
-                        bool keep_looking = true;
-                        while(keep_looking)
+                        if (mParticlesVector[free_particle].GetEraseFlag()==true)
                         {
-                            if (mParticlesVector[freeparticle].GetEraseFlag()==true)
+                            #pragma omp critical
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h:694]: Just pointing that this `critical` is killing your performance (not sure if the same behavior can be alternatively achieved).

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 15]
```diff
@@ -980,115 +755,78 @@ class MoveShallowWaterParticleUtility
...
-
-            BoundedMatrix<double, (3+2*TDim), 3 > pos; //7 particles (2D) or 9 particles (3D)
-            BoundedMatrix<double, (3+2*TDim), (TDim+1) > N;
+        BoundedMatrix<double, 3+2*TDim, 3> pos; //7 particles (2D) or 9 particles (3D)
+        BoundedMatrix<double, 3+2*TDim, TDim+1> N;
+        unsigned int free_particle = 0; //we start by the first position;
 
-            double mesh_scalar1;
-            array_1d<double,3> mesh_vector1;
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
        auto i_elem_begin = mrModelPart.ElementsBegin();
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 16]
```diff
@@ -980,115 +755,78 @@ class MoveShallowWaterParticleUtility
...
+        #pragma omp parallel for firstprivate(pos, N, free_particle)
+        for (int ii = 0; ii < static_cast<int>(mrModelPart.NumberOfElements()); ++ii) {
+            ModelPart::ElementsContainerType::iterator ielem = i_elem_begin + ii;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]:
```suggestion
            auto ielem = i_elem_begin + ii;
```

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 17]
```diff
@@ -980,115 +755,78 @@ class MoveShallowWaterParticleUtility
...
-            BoundedMatrix<double, (3+2*TDim), 3 > pos; //7 particles (2D) or 9 particles (3D)
-            BoundedMatrix<double, (3+2*TDim), (TDim+1) > N;
+        BoundedMatrix<double, 3+2*TDim, 3> pos; //7 particles (2D) or 9 particles (3D)
+        BoundedMatrix<double, 3+2*TDim, TDim+1> N;
+        unsigned int free_particle = 0; //we start by the first position;
 
-            double mesh_scalar1;
-            array_1d<double,3> mesh_vector1;
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
+        #pragma omp parallel for firstprivate(pos, N, free_particle)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: TLS.

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp [hunk 18]
```diff
@@ -100,109 +94,69 @@ void ShallowWaterUtilities::ComputeSmoothVelocity(ModelPart& rModelPart)
...
 void ShallowWaterUtilities::ComputeMomentum(ModelPart& rModelPart)
 {
-    #pragma omp parallel for
-    for (int i = 0; i < static_cast<int>(rModelPart.NumberOfNodes()); ++i)
-    {
-        auto it_node = rModelPart.NodesBegin() + i;
-        it_node->FastGetSolutionStepValue(MOMENTUM) = it_node->FastGetSolutionStepValue(VELOCITY) * it_node->FastGetSolutionStepValue(HEIGHT);
-    }
+    block_for_each(rModelPart.Nodes(), [&](NodeType& rNode){
+        rNode.FastGetSolutionStepValue(MOMENTUM) = rNode.FastGetSolutionStepValue(VELOCITY) * rNode.FastGetSolutionStepValue(HEIGHT);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp]:
```suggestion
        noalias(rNode.FastGetSolutionStepValue(MOMENTUM)) = rNode.FastGetSolutionStepValue(VELOCITY) * rNode.FastGetSolutionStepValue(HEIGHT);
```

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp [hunk 19]
```diff
@@ -214,111 +168,77 @@ void ShallowWaterUtilities::IdentifyWetDomain(ModelPart& rModelPart, Flags WetFl
...
-    {
-        auto it_node = rModelPart.NodesBegin() + i;
-        double& height = it_node->FastGetSolutionStepValue(HEIGHT);
+    block_for_each(rModelPart.Nodes(), [&](NodeType& rNode){
+        double& height = rNode.FastGetSolutionStepValue(HEIGHT);
         if (height < Thickness)
         {
             height = 0.5 * Thickness;
-            it_node->FastGetSolutionStepValue(MOMENTUM) = ZeroVector(3);
+            rNode.FastGetSolutionStepValue(MOMENTUM) = ZeroVector(3);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp]:
```suggestion
            noalias(rNode.FastGetSolutionStepValue(MOMENTUM)) = ZeroVector(3);
```

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp [hunk 20]
```diff
@@ -100,109 +94,69 @@ void ShallowWaterUtilities::ComputeSmoothVelocity(ModelPart& rModelPart)
...
         {
-            it_node->Set(SolidBoundaryFlag, true);
+            rNode.Set(SolidBoundaryFlag, true);
         }
         else
         {
-            auto topography_gradient = it_node->GetValue(TOPOGRAPHY_GRADIENT);
-            auto normal = it_node->FastGetSolutionStepValue(NORMAL);
+            auto topography_gradient = rNode.GetValue(TOPOGRAPHY_GRADIENT);
+            auto normal = rNode.FastGetSolutionStepValue(NORMAL);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp]:
```suggestion
            array_1d<double,3> normal = rNode.FastGetSolutionStepValue(NORMAL);
```
At some point we experienced some weird side effects when using `auto` for `array_1d<double,3>` types.

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp [hunk 21]
```diff
@@ -214,111 +168,77 @@ void ShallowWaterUtilities::IdentifyWetDomain(ModelPart& rModelPart, Flags WetFl
...
 }
 
 void ShallowWaterUtilities::SetMinimumValue(ModelPart& rModelPart, const Variable<double>& rVariable, double MinValue)
 {
-    #pragma omp parallel for
-    for (int i = 0; i < static_cast<int>(rModelPart.NumberOfNodes()); ++i)
-    {
-        auto& value = (rModelPart.NodesBegin() + i)->FastGetSolutionStepValue(rVariable);
+    block_for_each(rModelPart.Nodes(), [&](NodeType& rNode){
+        auto& value = rNode.FastGetSolutionStepValue(rVariable);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_utilities.cpp]:
```suggestion
        double& value = rNode.FastGetSolutionStepValue(rVariable);
```
For simple return types as `double` it's better not to use `auto`.

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 13 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: In that case, I need IndexPartition from parallel utilities. Does it support TLS?

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 5 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: I will take another look. There was a failing test related to that loop.

Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: done

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 7 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: changed to double

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 14 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h:694]: this is the bottleneck of the algorithm...

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 13 above)
Comment (rubenzorrilla) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: I think it does.

Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: done

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h (hunk 17 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h]: done

**Review by ipouplana — APPROVED** [review]
Given that you already reviewed all comments by Ruben. I approve this.

**Pull request merged**